File: flutter_app/lib/features/tasks/data/repositories/task_repository_impl.dart
import '../../domain/entities/task.dart';
import '../../domain/repositories/task_repository.dart';
import '../datasources/task_remote_datasource.dart';

class TaskRepositoryImpl implements TaskRepository {
  TaskRepositoryImpl({required TaskRemoteDataSource remoteDataSource})
      : _remoteDataSource = remoteDataSource;

  final TaskRemoteDataSource _remoteDataSource;

  List<Task> _tasks = [];
  bool _isLoaded = false;

  @override
  Future<List<Task>> getTasks() async {
    if (!_isLoaded) {
      final remoteTasks = await _remoteDataSource.fetchTasks();
      _tasks = remoteTasks.map((model) => model.toEntity()).toList();
      _isLoaded = true;
    }

    return List.unmodifiable(_tasks);
  }

  @override
  Future<Task> createTask(Task task) async {
    final createdTask = task.copyWith(id: _nextId());
    _tasks = [createdTask, ..._tasks];
    return createdTask;
  }

  @override
  Future<Task> updateTask(Task task) async {
    final index = _tasks.indexWhere((item) => item.id == task.id);

    if (index == -1) {
      throw StateError('Task not found.');
    }

    _tasks = List<Task>.from(_tasks);
    _tasks[index] = task;
    return task;
  }

  @override
  Future<void> deleteTask(int id) async {
    _tasks = _tasks.where((task) => task.id != id).toList();
  }

  int _nextId() {
    if (_tasks.isEmpty) {
      return 1;
    }

    final highestId = _tasks
        .map((task) => task.id)
        .reduce((current, next) => next > current ? next : current);

    return highestId + 1;
  }
}

File: RNX/src/features/tasks/data/repositories/TaskRepositoryImpl.ts
import type { Task, TaskInput } from '../../domain/entities/Task';
import type { TaskRepository } from '../../domain/repositories/TaskRepository';
import { TaskRemoteService } from '../services/TaskRemoteService';

export class TaskRepositoryImpl implements TaskRepository {
  private tasks: Task[] = [];
  private loaded = false;

  constructor(private readonly remoteService: TaskRemoteService) {}

  async getTasks(): Promise<Task[]> {
    if (!this.loaded) {
      const initialTasks = await this.remoteService.fetchTasks();

      this.tasks = initialTasks.map((task, index) =>
        this.remoteService.toEntity(task, index + 1),
      );
      this.loaded = true;
    }

    return [...this.tasks];
  }

  async createTask(task: TaskInput): Promise<Task> {
    const createdTask: Task = {
      id: this.nextId(),
      ...task,
    };

    this.tasks = [createdTask, ...this.tasks];

    return createdTask;
  }

  async updateTask(task: Task): Promise<Task> {
    const index = this.tasks.findIndex((currentTask) => currentTask.id === task.id);

    if (index === -1) {
      throw new Error('Task not found.');
    }

    this.tasks = this.tasks.map((currentTask) =>
      currentTask.id === task.id ? task : currentTask,
    );

    return task;
  }

  async deleteTask(id: number): Promise<void> {
    this.tasks = this.tasks.filter((task) => task.id !== id);
  }

  private nextId(): number {
    return this.tasks.length === 0
      ? 1
      : Math.max(...this.tasks.map((task) => task.id)) + 1;
  }
}